src/rtc.c

```diff
@@ -0,0 +1,792 @@
+    if( rtc_time->sec > 59 ) rtc_time->sec = 59;
+
+    /* Clamp the day of the month based on the year and month */
+    uint16_t year = rtc_time->year;
+    uint8_t month = rtc_time->month;
+    uint8_t days_in_month = DAYS_IN_MONTH[month];
+    bool leap_year = year % 400 == 0 || (year % 100 != 0 && year % 4 == 0);
+    if( month == 1 && leap_year ) days_in_month = 29;
+    if( rtc_time->day > days_in_month ) rtc_time->day = days_in_month;
+
+    /* Recalculate the day of the week based on calendar date */
+    month = month + 1; /* Adjust month to 1-indexed */
+    if( month < 3 )
+    {
+        month = month + 12;
+        year = year - 1;
+    }
+    rtc_time->week_day = (
+        rtc_time->day
+        + (2 * month)
+        + (6 * (month + 1) / 10)
+        + year
+        + (year / 4)
+        - (year / 100)
+        + (year / 400)
+        + 1
+    ) % 7;
+}
+
+/**
+ * @brief Read the current date/time from the real-time clock.
+ * 
+ * If the RTC is not detected or supported, this function will
+ * not modify the destination rtc_time parameter.
+ *
+ * Your code should call this once per frame to update the #rtc_time_t
+ * data structure. The RTC Subsystem maintains a cache of the
+ * most-recent RTC time that was read and will only perform an
+ * actual RTC read command if the cache is invalidated. The
+ * destination rtc_time parameter will be updated regardless of
+ * the cache validity.
+ * 
+ * Cache will invalidate every #RTC_GET_CACHE_INVALIDATE_TICKS.
+ * Cache will also invalidate when the tick counter overflows.
+ * Calling #rtc_set will also invalidate the cache.
+ *
+ * @param[out]  rtc_time
+ *              Destination pointer for the RTC time data structure
+ * 
+ * @return whether the rtc_time destination pointer data was modified
+ */
+bool rtc_get( rtc_time_t * rtc_time )
+{
+    /* libdragon currently only supports getting the time for Joybus RTC! */
+    if( rtc_present() != RTC_JOYBUS ) return false;
+
+    /* This should be overwritten the first time this function is called */
+    static rtc_time_t cache_time = { 2000, 0, 1, 0, 0, 0, 6 };
+
+    uint32_t current_ticks = get_ticks();
+    int32_t distance = TICKS_DISTANCE( current_ticks, rtc_get_cache_ticks );
+
+    if(
+        rtc_get_cache_ticks == 0 || /* cache manually invalidated */
+        distance < 0 || /* ticks counter overflow */
+        distance > RTC_GET_CACHE_INVALIDATE_TICKS 
+    )
+    {
+        joybus_rtc_read_time( &cache_time );
+        rtc_get_cache_ticks = current_ticks;
+    }
+    
+    memcpy( rtc_time, &cache_time, sizeof(rtc_time_t) );
+
+    return true;
+}
+
+/**
+ * @brief High-level convenience helper to set the RTC date/time.
+ *
+ * Prepares the RTC for writing, sets the new time, and resumes the clock.
+ *
+ * This process will take approximately 570 milliseconds to complete.
+ * Unfortunately, the best way to ensure that writes to the RTC have
+ * actually finished is by waiting for a fixed duration. Emulators may not
```

Comment:
This sentence is a little confusing to me. Does this mean that the user has to wait a fixed duration after calling `rtc_set`? I guess the meaning is that 570ms includes a conservative delay, but maybe it should be clarified.

Reply:
Fixed in https://github.com/DragonMinded/libdragon/pull/152/commits/544baa50639e55196c93c17c97b8cda1365f7bfc